### PR TITLE
video_core: Correct shader interpeter address register handling.

### PR DESCRIPTION
**Pull request by Steveice10**
Fixes two issues:
* The shader interpreter was applying the address register offset to all types of uniforms, when it is only applicable for float constant uniforms.
* The shader interpreter was not bounds checking the register index after adding the address register offset, which could cause a bad access crash.

Also fixed up some style warnings on a few flag checks.

- - -
This change is [Reviewable](https://reviewable.io/reviews/citra-emu/citra/6910)

### CODE REVIEW

**Review by SachinVin — APPROVED** [review]

**Comment by neobrain** [comment]
> The shader interpreter was applying the address register offset to all types of uniforms, when it is only applicable for float constant uniforms.

How do you know this is the correct behavior?

>The shader interpreter was not bounds checking the register index after adding the address register offset, which could cause a bad access crash. 

Does this happen in real-world workloads? This seems to indicate a bug elsewhere.

**Comment by Steveice10** [comment]
> > The shader interpreter was applying the address register offset to all types of uniforms, when it is only applicable for float constant uniforms.
> 
> How do you know this is the correct behavior?

https://www.3dbrew.org/w/index.php?title=GPU/Shader_Instruction_Set&mobileaction=toggle_view_desktop#Relative_addressing

> > The shader interpreter was not bounds checking the register index after adding the address register offset, which could cause a bad access crash.
> 
> Does this happen in real-world workloads? This seems to indicate a bug elsewhere.

It’s documented that the address registers take negative values. Regardless of whether the game is exhibiting this behavior for other reasons, it needs to be handled.

**Comment by Steveice10** [comment]
I've also checked on hardware by:
* Filling float uniforms with different color vectors.
* Moving a negative float vector into a0.
* Using the float uniform indexed by a0 as the color.

As a result I got (0, 0, 0, 0) as the color rendered.

**Comment by Steveice10** [comment]
I've performed a more thorough hardware test on a range of values and arrived at the following behavior:
* If the value is out of byte bounds, the result is (0, 0, 0, 0).
* Given the value without its sign bit:
  * If the value is >= 96, the result is (1, 1, 1, 1).
  * Otherwise, the uniform value is returned.

I've updated the code to match this behavior.

Hardware Test: 
[mova-bounds-test.zip](https://github.com/citra-emu/citra/files/12428522/mova-bounds-test.zip)

Press up to increment and down to decrement the register.

File: src/video_core/shader/shader_interpreter.cpp [hunk 1]
```diff
@@ -130,19 +133,36 @@ static void RunInterpreter(const ShaderSetup& setup, UnitState& state, DebugData
...
             case RegisterType::Temporary:
-                return &state.registers.temporary[source_reg.GetIndex()].x;
+                return &state.registers.temporary[index].x;
 
             case RegisterType::FloatUniform:
-                return &uniforms.f[source_reg.GetIndex()].x;
+                if (address_register_index != 0) {
+                    index += state.address_registers[address_register_index - 1];
+                    // If the index is out of s8 bounds, the result is all zero.
+                    if (index < std::numeric_limits<s8>::min() ||
```
Comment (GPUCode) [inline src/video_core/shader/shader_interpreter.cpp:150]: By doing ```index = std::abs(index)``` you could check only for > 128 and >= 96 cases

File: src/video_core/shader/shader_interpreter.cpp [hunk 2]
```diff
@@ -113,7 +113,10 @@ static void RunInterpreter(const ShaderSetup& setup, UnitState& state, DebugData
     const auto& program_code = setup.program_code;
 
     // Placeholder for invalid inputs
-    static f24 dummy_vec4_float24[4];
+    static f24 dummy_vec4_float24_zeros[4] = {f24::FromFloat32(0.0f), f24::FromFloat32(0.0f),
```
Comment (GPUCode) [inline src/video_core/shader/shader_interpreter.cpp:116]: Can use f24::Zero() and f24::One() here for simplicity

File: src/video_core/shader/shader_interpreter.cpp (hunk 1 above)
Comment (SachinVin) [reply src/video_core/shader/shader_interpreter.cpp:150]: Seems like unnecessary optimisation at the expense of readability

**Comment by Steveice10** [comment]
Did another round of hardware testing with inverted color order and a different base register and have updated the behavior:
* If the offset is out of signed byte bounds, the offset is 0.
* The offset is added to the register index and the result is masked with 0x7F.
* If the register index is greater than 95 (out of bounds register number), (1, 1, 1, 1) is returned.

Updated tests, with instructions included as well now: 
[mova-bounds-test.zip](https://github.com/citra-emu/citra/files/12432082/mova-bounds-test.zip)

**Comment by GPUCode** [comment]
For anyone interested, this is another test based on steveice's test above but readbacks the colour values which make it a little easier to see the value fluctuation.
[simple_tri.zip](https://github.com/citra-emu/citra/files/12433319/simple_tri.zip)

**Pull request merged**